### DISCUSSION TITLE
Flux bootstrap for another cluster in another branch

### DISCUSSION THREAD

**Discussion by sachinshakya507** [fluxcd/flux2#3460] [General]
Let say i have two cluster A and B.
Previously what i can do is flux bootstrap same repository on two different branches let say  branch main for cluster A and  branch master for cluster B.
But now i cannot have this feature.

**Comment by kingdonb** [comment]
I'm not sure what is suggested here. You can bootstrap two clusters in the same repo. They can either be on the same branch, or on different branches, (but they cannot have the same `bootstrap .. --path` in the same branch.)

What version of Flux are you using, and is this a regression from a previous version? Can you provide steps to reproduce and show what you mean please.

**Comment by sachinshakya507** [comment]
I am using flux version =flux version 0.38.0
When i use flux bootstrap for first time for cluster A with any branch it works.
flux bootstrap github \                  
    --owner=username \
    --repository=repo \
    --branch=main \
    --personal \
    --path=app

But when i try for another cluster B with any other branch than main then it not works.
flux bootstrap github \                  
    --owner=usernmame\
    --repository=repo\
    --branch=other_branch \
    --personal \
    --path=app

Output from flux bootstrap
► cloning branch "other_branch" from Git repository "https://github.com/username/repo.git"
⚠️  clone failure: unable to clone 'https://github.com/username/repo.git': couldn't find remote ref "refs/heads/other_branch"
⚠️  clone failure: unable to clone 'https://github.com/username/repo.git': repository already exists
✗ failed to clone repository: unable to clone 'https://github.com/username/repo.git': repository already exists

**Reply by xunholy** [reply]
@sachinshakya507 are you attempting to bootstrap two separate branches into the same cluster?

**Reply by kingdonb** [reply]
You cannot bootstrap the same path into two clusters from the same branch, but you should be able to bootstrap the same path from different branches into two different clusters AFAIK.

I think what's happening is, the bootstrap process is trying to create the repo, or clone it. There is no conditional behavior for "the repo already exists but the branch I want does not exist".

If you create the branch ahead of attempting the second bootstrap, and make sure the branch ref has been pushed, does that help?

Edit: This is maybe a bug, but I'm not sure what the correct behavior would be if you target bootstrap for a branch that does not exist. Do you want a new branch with no parent? Or do you want a branch to fork from the existing main branch? Flux has no way of knowing which you want, so I hesitate to call this a bug.

You should create the branch ahead of the second bootstrap with the target ref you want your new branch to start from, unless it's a new repo (then Flux can create the branch and repo out of whole cloth, since that's the usual bootstrap flow.)

**Comment by maeghan-porter** [comment]
On a similar vein, we have our flux setup in such a way where our clusters have overlays that use `flux-system` as a base. But when trying to upgrade flux, I'm getting an error because the cluster I'm trying the upgrade on points to an overlay folder, not the `flux-system` folder, but I need the updates to run agains the base `flux-system` folder. Any advice?
